ntls-io/ntc-web
# Full SDK with enclave signing service

### Description

This branch integrates the functions needed for the NTC vault users to create and interact with data pools/DRTs. These functions are represented by methods that encompass the algorand-js sdk and the signing service of our enclave. These methods in this PR include all transactions needing a box reference in their transaction array. 
Tests are built in using the jasmine karma framework. 

The following methods were added and/or updated in this PR.
- Create Data Pool method, this includes 5 transactions to create a data pool. One of them is an enclave transaction but a DEMO transaction is used to mimic the behaviour of an enclave transaction until there is the enclave transactions is functional. 
- Create DRT method, this includes 2 transactions - create DRT txn and store DRT txn
- Buy DRT method, this includes 3 transactions - Optin to DRT txn, and a group transaction consisting of a payment txn nd instruction to buy txn. 
- De-list and List DRT Methods, this includes 1 txn each. 
- Join Data Pool as pending contributor method, this includes the buy DRT method to purchase the append DRT and a 3 transaction group - append asset transfer txn, execution fee txn, and instruction txn.
- Claim contributor token Method, this includes 2 transactions - contributor asset optin txn, claim contributor token instruction txn. 
- Claim royalties as a data contributor method, this includes 1 transaction - claim royalty instruction txn.
- Redeem DRT method, this includes 3 transactions in a group transaction - asset transfer txn, execution fee txn, redeem drt instruction txn. 

Tests were added for each method including a test for Nautilus to claim our fees from the data pool, and a DEMO enclave contributor approval transaction to confirm a pending contributor to a confirmed contributor. 


### How to Test
The tests were conducted by launching a data pool smart contract on the algorand testnetwork and interacting with the data pool through test accounts. These test accounts are defined in the sdk/createPool/fullPoolCreateAndOpertaions.spec.ts testing file. I have dispensed enough algos into each account to last for a while but after time we will need to dispense more funds into each test account. 

`npm test`

Open your terminal to see logs of all the blockchain transaction tests.
